### PR TITLE
Tune/memoize Mailchimp interest categories/groups

### PR DESCRIPTION
**Pull request by jayzes**
This call seems to be taking up to 30s to come back, which is sort of bonkers. I’m hopeful that memoizing it will help it not be quite so insane, though it means that we’ll have to shut down the reader before making interest group changes if we want to do so. Seems fine (this code would likely need to change in that case anyway) but worth noting.

Example New Relic transaction trace:

screen shot 2016-12-09 at 2 34 33 pm

https://rpm.newrelic.com/accounts/616593/applications/28688571/transactions#id=5b224f746865725472616e73616374696f6e2f4261636b67726f756e642f456c6c6f3a3a4b696e65736973436f6e73756d65723a3a4d61696c6368696d7050726f636573736f722f757365725f6368616e6765645f737562736372697074696f6e5f707265666572656e636573222c22225d&app_trace_id=c7e540dd-be53-11e6-bd53-c81f66b8ceca_17302_17926

### CODE REVIEW

File: lib/ello/mailchimp_wrapper.rb [hunk 1]
```diff
@@ -49,54 +49,50 @@ def subscriber_hash(email)
...
-  def users_list_categories_interest_group
-    id = retrieve_category_id('Categories')
-    users_list.interest_categories(id).interests.retrieve['interests'].each_with_object({}) do |category, categories|
-      categories[category['name'].downcase] = category['id']
+  def users_list_categories_interest_group_interests
+    @users_list_categories_interest_group_interests ||= begin
+      id = find_category_id_from_name('Categories')
+      users_list.interest_categories(id).interests.retrieve['interests'].each_with_object({}) do |category, categories|
+        categories[category['name'].downcase] = category['id']
+      end
```
Comment (jayzes) [inline lib/ello/mailchimp_wrapper.rb]: No need to apologize at all - I changed a few names to try to make it a bit more understandable for myself, but I don't think it was poorly structured to start. Always tough when two domains collide with the same terminology.

That's solid insight on the `map_category_names_to_ids` expectation, and definitely something I didn't think about. I'm thinking about extracting an explicit read-through cache object out of there to hold onto the Mailchimp category mappings, which we could then check explicitly in `map_category_names_to_ids` and force a refresh if we get a user preference that doesn't appear in the cache. That would let us effectively memoize the request most of the time, but deal more gracefully with a miss. Do you think that would solve the underlying issue? We could do the same in the other use case (preferences) as well.

**Comment by jayzes** [comment]
@Justin-Holmes I think I did this in a simpler way that ultimately accomplishes the same thing, but would love one more look to make sure i'm not crazy.

Also, i'm using a class variable which is normally a little funky with threads but I think will be fine here since we're not directly mutating that cached value, just ensuring that it gets reloaded.

File: lib/ello/mailchimp_wrapper.rb [hunk 2]
```diff
@@ -100,14 +104,17 @@ def map_category_names_to_ids(prefs)
   end
 
   def assign_category_prefs(categories, prefs)
-    categories.each do |category|
-      unless prefs.key?(category)
-        id = find_category_id_from_name('Categories')
-        users_list.interest_categories(id).interests.create(body: { name: category.capitalize })
+    prefs.dup.tap do |p|
```
Comment (Justin-Holmes) [inline lib/ello/mailchimp_wrapper.rb]: Ah, good call on tap here.

File: lib/ello/mailchimp_wrapper.rb [hunk 3]
```diff
@@ -100,14 +104,17 @@ def map_category_names_to_ids(prefs)
...
+      categories.each do |category|
+        unless p.key?(category)
+          id = find_category_id_from_name('Categories')
+          users_list.interest_categories(id).interests.create(body: { name: category.capitalize })
+          # Invalidate the memoized attribute so map_category_names_to_ids re-fetches the groups
+          fetch_users_list_categories_interest_group_interests!
```
Comment (Justin-Holmes) [inline lib/ello/mailchimp_wrapper.rb]: Nice, this seems much simpler than going the cache object route.

**Comment by Justin-Holmes** [comment]
This LGTM! I guess we'll see if there is any weirdness with the class variable. Hopefully, there won't be any weird race conditions if multiple users try to follow a non-existent category at the same time.

**Comment by jayzes** [comment]
Totally! My thinking is that either way we should be fine, i.e. it's a race where we don't care who the winner is because any of them should have the new data.

**Pull request merged**